lib/open_food_network/subscription_service.rb

```diff
@@ -0,0 +1,25 @@
+module OpenFoodNetwork
+  class SubscriptionService
+    # Includes the following variants:
+    # - Variants of permitted producers
+    # - Variants of hub
+    # - Variants that are in outgoing exchanges where the hub is receiver
+    def self.available_variants(distributor)
+      permitted_order_cycle_enterprise_ids = EnterpriseRelationship.permitting(distributor).with_permission(:add_to_order_cycle).pluck(:parent_id)
+      permitted_producer_ids = Enterprise.is_primary_producer.where('enterprises.id IN (?)', permitted_order_cycle_enterprise_ids).pluck(:id)
+
+      outgoing_exchange_variant_ids = ExchangeVariant.select("DISTINCT exchange_variants.variant_id")
+        .joins(:exchange)
+        .where(exchanges: { incoming: false, receiver_id: distributor.id })
+        .pluck(:variant_id)
+
+      variant_conditions = ["spree_products.supplier_id IN (?)", permitted_producer_ids | [distributor.id]]
+      if outgoing_exchange_variant_ids.present?
+        variant_conditions[0] << " OR spree_variants.id IN (?)"
+        variant_conditions << outgoing_exchange_variant_ids
+      end
+
+      Spree::Variant.joins(:product).where(is_master: false).where(*variant_conditions)
```

Comment:
This has been done in this commit: https://github.com/openfoodfoundation/openfoodnetwork/pull/2733/commits/9b55e08e341129e67fe3dcf6a73e7eaac683514a I also made a slight SQL query optimization in `self.permitted_producer_ids`.

Next step for this would be to make these methods instance methods and have an instance variable `@distributor`, so we don't have to pass it around class methods. Issue is here: #3363